### ISSUE TITLE
getdiag() for GMats refers to mkdiag()

### ISSUE THREAD

**Issue by DanielTakeshi** [BIDData/BIDMat#18]
Version: master branch, after commit 97cda3d on Feb 6 2015.

The code for MatFunctions.scala has:

def getdiag(a:DMat) = DMat(a.getdiag)
def getdiag(a:FMat) = FMat(a.getdiag)
def getdiag(a:IMat) = IMat(a.getdiag)
def getdiag(a:CMat) = CMat(a.getdiag)
def getdiag(a:GMat) = a.mkdiag

I am not sure why there's a mkdiag for the GPU case. I can get around this by calling it from the GMat code, i.e., a.getdiag() if a is a GMat.

Some examples:

scala> val a = 1\2\3 on 4\5\6 on 7\8\9
a: BIDMat.IMat =
   1   2   3
   4   5   6
   7   8   9

scala> getdiag(a)
res8: BIDMat.IMat =
   1
   5
   9

scala> getdiag(BIDMat.GMat(a))
java.lang.RuntimeException: mkdiag requires a vector argument, but dims= 3 3
  at BIDMat.GMat.mkdiag(GMat.scala:636)
  at BIDMat.MatFunctions$.getdiag(MatFunctions.scala:1437)
  ... 33 elided

**Comment by jcanny** [comment]
Looks like its just a bug. Feel free to fix it. btw, can you check if 
there is a getdiag for GIMat GDMat and GLMat, and if so, add those lines 
in MatFunctions?

-john

**Comment by jcanny** [comment]
Hi Daniel,
I could use some help with BIDMach's documentation. I'm planning a 1.0 
release to correspond with a Strata conference talk in a week and a 
half. You mentioned you enjoy documentation - it would really help if 
you could write some. I realize you dont know what's going on in many of 
the fucntions - that's fine. Actually if you can point me at the places 
where you think documentation is most needed, that would really help so 
I can focus my own efforts. We have incomplete API docs, quick reference 
(but only for BIDMat), and some general topical documentation on the 
wiki. There are also several iscala tutorials. I think the tutorials 
need to be simplified - we should start with how to create a leaner and 
run "train" and "predict". and then there are various sample scripts in 
the scripts directory.  Let me know your thoughts on what would help most.

-john

**Comment by DanielTakeshi** [comment]
John,

There's also an issue with mkdiag for GMats that I fixed:

scala> val diag = BIDMat.GMat(1 on 2 on 3)
diag: BIDMat.GMat =
   1
   2
   3

scala> mkdiag(diag)
java.lang.RuntimeException: mkdiag requires a vector argument, but dims= 3 1
  at BIDMat.GMat.mkdiag(GMat.scala:643)
  at BIDMat.MatFunctions$.mkdiag(MatFunctions.scala:1446)
  ... 33 elided

The issue was with the GMat's mkdiag method, where the math.max should be math.min (same thing in GDMat).

def mkdiag():GMat = {
if (math.max(nrows, ncols) != 1) throw new RuntimeException("mkdiag requires a vector argument, but dims= %d %d" format (nrows, ncols))

Other fixes:

(2) Changed that GMat's mkdiag to getdiag
(3) LMat matrices were missing mkdiag and getdiag completely, so I assume those should be there (it's just longs, after all)
(4) Added GIMat, GLMat, and GDMat to getdiag and mkdiag
(5) Any other types to add? Right now there are 9 supported types for getdiag/mkdiag.

One thing about the GPU matrices and getdiag/mkdiag: the GDMat had its own mkdiag/getdiag but for some reason the GIMats and GLMats didn't. I just copied the GDMat's method over and changed the cases where it said GDMat to GIMat or GLMats. My tests ran into issues with cuda memory allocation so I also made sure to use the appropriate Sizeof.INT, Sizeof.LONG, etc.

----------

With regards to documentation, yes I am definitely happy to help out on that. One thing is that I still need to get Independent Component Analysis implemented *well* enough. My FastICA imlementation isn't quite there yet...I'll keep you posted on that this weekend.
